File: apple/Sources/JanusGateway/JaHandle.swift
//
//  JaHandle.swift
//
//
//  Created by Hamza Jadid on 16/09/2024.
//

import Foundation
import UniFFI

/// General purpose plugin handle
public final class JaHandle {
    let handle: Handle
    public var delegate: JaHandleDelegate?
    private var continuation: AsyncStream<JaHandleEvent>.Continuation?

    /// Get an async stream of incoming Janus events for this handle
    ///
    /// - Returns: An async stream of incoming events
    public var events: AsyncStream<JaHandleEvent> {
        get async {
            await handle.startEventLoop(cb: self)

            return AsyncStream { continuation in
                self.continuation = continuation
            }
        }
    }

    init(handle: Handle) {
        self.handle = handle
    }

    /// Sends a message without waiting for any response or acknowledgment
    ///
    /// - Parameters:
    ///     - data: Data to be sent
    ///     - jsep: JavaScript Session Establishment Protocol
    public func fireAndForget(data: Data, jsep: Jsep? = nil) async throws {
        if let jsep {
            try await handle.fireAndForgetWithJsep(data: data, jsep: jsep)
        } else {
            try await handle.fireAndForget(data: data)
        }
    }

    /// Sends a message and waits until the server acknowledges or timeout
    ///
    /// - Parameters:
    ///     - data: Data to be sent
    ///     - timeout: The maximum amount of time to wait on an acknowledgment before we consider
    ///     the request as failed or times out.
    public func sendWaitonAck(
        data: Data, timeout: TimeInterval
    ) async throws {
        try await handle.sendWaitonAck(data: data, timeout: timeout)
    }

    /// Sends a message and waits until the server returns a response or timeout
    ///
    /// - Parameters:
    ///     - data: Data to be sent
    ///     - timeout: The maximum amount of time to wait on a response before we consider the
    ///     request as failed or times out.
    public func sendWaitonResult(
        data: Data,
        timeout: TimeInterval
    ) async throws -> Data {
        try await handle.sendWaitonResult(data: data, timeout: timeout)
    }

    /// Hang up the associated PeerConnection but keep the handle alive
    ///
    /// - Parameters:
    ///     - timeout: The maximum amount of time to wait on a response before we consider the
    ///     request as failed or times out.
    public func hangup(timeout: TimeInterval) async throws {
        try await handle.hangup(timeout: timeout)
    }

    /// Destroy the plugin handle
    ///
    /// - Parameters:
    ///     - timeout: The maximum amount of time to wait on a response before we consider the
    ///     request as failed or times out.
    public func detach(timeout: TimeInterval) async throws {
        try await handle.detach(timeout: timeout)
    }

    public func trickleSingleCandidate(
        candidate: Candidate, timeout: TimeInterval
    ) async throws {
        try await handle
            .trickleSingleCandidate(candidate: candidate, timeout: timeout)
    }

    public func trickleCandidates(
        candidates: [Candidate], timeout: TimeInterval
    ) async throws {
        try await handle
            .trickleCandidates(candidates: candidates, timeout: timeout)
    }

    public func completeTrickle(timeout: TimeInterval) async throws {
        try await handle.completeTrickle(timeout: timeout)
    }
}

extension JaHandle: HandleCallback {
    public func onPluginEvent(event: Data) {
        delegate?.didReceiveHandleEvent(event: .plugin(event))
        continuation?.yield(.plugin(event))
    }

    public func onHandleEvent(event: GenericEvent) {
        delegate?.didReceiveHandleEvent(event: .handle(event))
        continuation?.yield(.handle(event))
    }
}

File: android/core/src/main/java/com/ghamza/janus/core/JaHandle.kt
package com.ghamza.janus.core

import com.ghamza.janus.bindings.Handle
import com.ghamza.janus.bindings.HandleCallback
import com.ghamza.janus.bindings.Jsep
import com.ghamza.janus.bindings.GenericEvent
import kotlinx.coroutines.channels.ProducerScope
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.callbackFlow
import java.time.Duration

class JaHandle(val handle: Handle): HandleCallback {
    private var events: ProducerScope<String>? = null

    fun stream() = callbackFlow {
        events = this
        handle.startEventLoop(this@JaHandle)
        awaitClose { }
    }

    suspend fun fireAndForget(msg: String) {
        handle.fireAndForget(msg.toByteArray())
    }

    suspend fun fireAndForget(msg: String, jsep: Jsep) {
        handle.fireAndForgetWithJsep(msg.toByteArray(), jsep)
    }

    suspend fun sendWaitOnAck(msg: String, timeout: Duration) {
        handle.sendWaitonAck(msg.toByteArray(), timeout)
    }

    suspend fun sendWaitOnResult(msg: String, timeout: Duration): String {
        val bytes = handle.sendWaitonResult(msg.toByteArray(), timeout)
        return String(bytes)
    }

    override fun onPluginEvent(event: ByteArray) {
        events?.trySend(String(event))
    }
    
    override fun onHandleEvent(event: GenericEvent) {
        
    }
}
